Allow configuration of OpenAPI output version

Azquelt opened the pull request:
- [x] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [x] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [x] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Allow the version of the OpenAPI spec used for the output of `/openapi` to be configured.

`mpOpenAPI-4.0` supports 3.1 and 3.0, earlier versions only support 3.0

The user can configure either:
- a three-digit version (e.g. `3.0.3`) which is used as-is in the output document
- a two-digit version (e.g. `3.0`) which is replaced with a three-digit version in the output document

To change the output format, we only have to change the version number included in the model before its serialized. 

In `mpOpenAPI-4.0`, the model always corresponds to the OpenAPI 3.1 spec. Smallrye OpenAPI takes care of transforming to 3.0 during serialization if required.

Resolves #28859

Built on #29769, all but the last few commits will be removed when that PR is merged.

Azquelt committed:
Allow configuration of OpenAPI output version

For mpOpenAPI-4.0, this allows the user to switch between producing
OpenAPI 3.1 and 3.0.

Azquelt committed:
Update to smallrye openapi 4.0.0-alpha6

Azquelt committed:
Update MP OpenAPI project settings

Azquelt committed:
Add OpenAPI version config tests

Test that the OpenAPI spec version can be configured for the mpOpenAPI
feature.

Azquelt committed:
Add openAPIVersion to test config model

This will be used to dynamically update the config during FAT tests.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/OpenAPIVersionConfigImpl.java:
```diff
@@ -0,0 +1,282 @@
+        if (Objects.equals(versionString, configuredVersionString)) {
+            // Our config has not changed, return immediately to avoid potentially logging warnings
+            if (TraceComponent.isAnyTracingEnabled() && tc.isEventEnabled()) {
+                Tr.event(this, tc, "Config unchanged", versionString);
```
An edge case. Lets say the customer sets it to null. On activate configuredVersion will be set to defaultVersion(). Then on modified null != defaultVersion.

The only effect is we'll get the wrong event message "no OpenAPI version set" rather than "Config unchanged".

Azquelt replied:
Not quite.

If we read the property as `null`, we'll set `configuredVersion` to the `defaultVersion()` and `configuredVersionString` to `null`.

The equality check is done against `configuredVersionString`.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/OpenAPIVersionConfigImpl.java:
```diff
@@ -0,0 +1,282 @@
+    }
+
+    @Override
+    public void applyConfig(OpenAPI model) {
```
This method does not beta guard correctly. It doesn't throw an UnsupportedOperationException if we're not beta, issue a warning message, and isn't marked deprecated.

Azquelt replied:
Will check, I think throwing UnsupportedOperationException and marking deprectated is only needed when we're adding methods to API or SPI classes. Since the user has access to that class, they could call that method in a non-beta version, at which point we want to throw an exception.

In contrast, this method does get called for both beta and production versions, it just does nothing in the production version.

Azquelt replied:
Those rules are indeed [only for API/SPI methods](https://github.com/OpenLiberty/open-liberty/wiki/Beta-Fencing#beta-fencing-using-jvm-system-property).

However, I did notice while checking the beta fencing that I haven't updated the metatype :facepalm:

Azquelt replied:
I've now updated the metatype.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/OpenAPIVersionConfigImpl.java:
```diff
@@ -0,0 +1,282 @@
+    protected boolean requiresReplacement(OpenAPIVersion modelVersion) {
+        if (modelVersion.major == configuredVersion.major
+            && modelVersion.minor == configuredVersion.minor
+            && modelVersion.patch >= 0
```
This if statement will always fall into the else if modelVersion.patch is -1 , even if both modelVersion and configueredVersion are completely identical. I want to confirm this is correct.

Azquelt replied:
This is correct. The modelVersion [must be three digits](https://spec.openapis.org/oas/v3.1.0.html#versions) whereas the configured version can be two.

I'll break this condition up and add comments.

benjamin-confino commented on dev/io.openliberty.microprofile.openapi.4.0.internal.services/src/io/openliberty/microprofile/openapi40/internal/services/impl/OpenAPIVersionConfig40Impl.java:
```diff
@@ -0,0 +1,54 @@
+    @Override
+    protected String getReplacementVersion() {
+        // If the user has configured a two-digit version, convert to the corresponding three-digit version
+        if (configuredVersion.equals(VERSION_30)) {
```
Its a small thing but I'd suggest having a map in the superclass and the subclasses ensure it has the correct contents rather than copying and pasting this code.

Azquelt replied:
I'm having a think about this.

When we expect a class to be extended, it's much simpler to require the extender to implement certain methods, rather than expecting them to manipulate the state of the parent object.

Azquelt replied:
I've left the code as-is but updated the javadoc to better explain what's expected of subclasses.

I'm still a bit tempted to abstract a completely version-agnostic abstract class and hide things that subclasses shouldn't touch.

Azquelt commented:
#libby
#build

Azquelt commented:
FAO ID/L2 reviewer:

This PR has changes from #29769 which have already been reviewed.

The new NLS changes are in the more recent commit here: https://github.com/OpenLiberty/open-liberty/pull/29877/commits/ad1ac1eaa57cda253b8fda02584c28f1cceefbd9#diff-53fa1abc4002268be3f730d1c6de09a39bd57b4f450847224b06f24d3ca8f0ae

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+
+# {1} - the server.xml element name, either includeApplication or excludeApplication
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W=CWWKO1678W: The configured {0} application name in the {1} element is not valid because it contains the "/" character.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=An application cannot name cannot include a "/" as "/" is used to separate the application from its component modules.
```
```suggestion
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=An application cannot name cannot include a "/" as the "/" is used to separate the application from its component modules.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+# {1} - the server.xml element name, either includeApplication or excludeApplication
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W=CWWKO1678W: The configured {0} application name in the {1} element is not valid because it contains the "/" character.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.explanation=An application cannot name cannot include a "/" as "/" is used to separate the application from its component modules.
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=Remove the "/" from the includeApplication or excludeApplication element in your server.xml, or convert that element to includeModule or excludeModule to reference a single module.
```
```suggestion
OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=Remove the "/" from the includeApplication or excludeApplication element in your server.xml file, or convert that element to includeModule or excludeModule to reference a single module.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPEN_API_SLASH_IN_APPLICATION_CWWKO1678W.useraction=Remove the "/" from the includeApplication or excludeApplication element in your server.xml, or convert that element to includeModule or excludeModule to reference a single module.
+
+# {1} - the server.xml element name, either includeModule or excludeModule
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid it is not in the form "<application-name>/<module-name>".
```
```suggestion
OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid as it is not in the format "<application-name>/<module-name>".
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+
+# {1} - the server.xml element name, either includeModule or excludeModule
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid it is not in the form "<application-name>/<module-name>".
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
```
```suggestion
OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=A module must be in the format "<application-name>/<module-name>"; therefore, a module name without a "/" is not valid.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+# {1} - the server.xml element name, either includeModule or excludeModule
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W=CWWKO1679W: The configured {0} module name in the {1} element is not valid it is not in the form "<application-name>/<module-name>".
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication to reference a whole application.
```
```suggestion
OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=Correct the includeModule or excludeModule element in your server.xml file, or convert that element to includeApplication or excludeApplication to reference an entire application.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.explanation=An module must be in the format <application-name>/<module-name>, therefore a module name without a "/" is invalid.
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication to reference a whole application.
+
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W=CWWKO1680W: The {1} application name in the {0} configuration element does not match the name of any deployed application but it does match either the name from the deployment descriptor or the archive name of the {2} application. The application name used here must be the application name specified in server.xml, or the archive file name with the extens ...
```
```suggestion
OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W=CWWKO1680W: The {1} application name in the {0} configuration element does not match the name of any deployed application, but it matches either the name from the deployment descriptor or the archive name of the {2} application. The application name that is used here must be the one specified in the server.xml file, or if no name is specified, the archive file name without the extension.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPEN_API_SLASH_IN_MODULE_CWWKO1679W.useraction=Correct the includeModule or excludeModule element in your server.xml, or convert that element to includeApplication or excludeApplication to reference a whole application.
+
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W=CWWKO1680W: The {1} application name in the {0} configuration element does not match the name of any deployed application but it does match either the name from the deployment descriptor or the archive name of the {2} application. The application name used here must be the application name specified in server.xml, or the archive file name with the extens ...
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.explanation=The application name reference in the mpOpenAPI configuration element matches the name from an application deployment descriptor, but does not match any application name defined in server.xml.
```
```suggestion
OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.explanation=The application name reference in the mpOpenAPI configuration element matches the name from an application deployment descriptor, but does not match any application name that is defined in the server.xml file.
```

benjamin-confino commented:
@ramkumar-k-9286 OpenAPI.nlsprops was updated in my PR at https://github.com/OpenLiberty/open-liberty/pull/29769

Its also shows up because my PR is a prerequisite for Andrew's so he's included my commits. I've made the changes in that branch.

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.explanation=The application name reference in the mpOpenAPI configuration element matches the name from an application deployment descriptor, but does not match any application name defined in server.xml.
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.useraction=Check that all applications started correctly and that the names in the configuration element are specified correctly.
+
+OPENAPI_VERSION_INVALID_CWWK06181W=CWWKO1681W: The configured {0} OpenAPI version is not in a valid format, it must be either two or three numbers separated by the period (.) character. The supported values are {1}. The default version of {2} will be used instead.
```
```suggestion
OPENAPI_VERSION_INVALID_CWWK06181W=CWWKO1681W: The configured {0} OpenAPI version is not in a valid format, it must be either two or three numbers that are separated by a period (.) character. The supported values are {1}. The default version of {2} is used instead.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPENAPI_USING_WRONG_NAME_SOURCE_CWWKO1680W.useraction=Check that all applications started correctly and that the names in the configuration element are specified correctly.
+
+OPENAPI_VERSION_INVALID_CWWK06181W=CWWKO1681W: The configured {0} OpenAPI version is not in a valid format, it must be either two or three numbers separated by the period (.) character. The supported values are {1}. The default version of {2} will be used instead.
+OPENAPI_VERSION_INVALID_CWWK06181W.explanation=The openAPIVersion attribute of the mpOpenAPI configuration element is set to a value which is not a valid OpenAPI version.
```
```suggestion
OPENAPI_VERSION_INVALID_CWWK06181W.explanation=The openAPIVersion attribute of the mpOpenAPI configuration element is set to a value that is not a valid OpenAPI version.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPENAPI_VERSION_INVALID_CWWK06181W.explanation=The openAPIVersion attribute of the mpOpenAPI configuration element is set to a value which is not a valid OpenAPI version.
+OPENAPI_VERSION_INVALID_CWWK06181W.useraction=Update the openAPIVersion attribute to one of the supported values from the error message.
+
+OPENAPI_VERSION_NOT_SUPPORTED_CWWK06182W=CWWKO1682W: The configured {0} OpenAPI version is not supported by this version of mpOpenAPI. The supported values are {1}. The default version of {2} will be used instead.
```
```suggestion
OPENAPI_VERSION_NOT_SUPPORTED_CWWK06182W=CWWKO1682W: This version of mpOpenAPI does not support the configured {0} OpenAPI version. The supported values are {1}. The default version of {2} is used instead.
```

ramkumar-k-9286 commented on dev/io.openliberty.microprofile.openapi.internal.common/resources/io/openliberty/microprofile/openapi/internal/resources/OpenAPI.nlsprops:
```diff
@@ -130,3 +130,25 @@ OPEN_API_DOC_PATH_INVALID_CWWKO1676E.useraction=Ensure the OpenAPI Path contains
+OPENAPI_VERSION_INVALID_CWWK06181W.useraction=Update the openAPIVersion attribute to one of the supported values from the error message.
+
+OPENAPI_VERSION_NOT_SUPPORTED_CWWK06182W=CWWKO1682W: The configured {0} OpenAPI version is not supported by this version of mpOpenAPI. The supported values are {1}. The default version of {2} will be used instead.
+OPENAPI_VERSION_NOT_SUPPORTED_CWWK06182W.explanation=The openAPIVersion attribute of the mpOpenAPI configuration element is set to an OpenAPI version which is not supported by the running version of the mpOpenAPI feature.
```
```suggestion
OPENAPI_VERSION_NOT_SUPPORTED_CWWK06182W.explanation=The openAPIVersion attribute of the mpOpenAPI configuration element is set to an OpenAPI version that is not supported by the running version of the mpOpenAPI feature.
```

Azquelt committed:
ID corrections for OpenAPI messages

From ID review

Co-authored-by: Ramkumar <88021034+ramkumar-k-9286@users.noreply.github.com>

Azquelt commented:
Thank you @ramkumar-k-9286, all your suggestions have been applied.

Azquelt commented:
#libby
#build

Azquelt commented:
#libby

Azquelt commented:
#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

benjamin-confino reviewed (dismissed):
Only changes since last review were ID changes, which are approved by ID.

Azquelt commented:
Rebased on updated #29769

#libby
#build
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

Azquelt commented:
Personal build looks good, this is ready to merge after it's been re-reviewed. The only change since last re-review was to rebase on #29769 to pick up a test fix.

tevans78 commented on dev/fattest.simplicity/src/com/ibm/websphere/simplicity/config/MpOpenAPIElement.java:
```diff
@@ -21,8 +21,9 @@
 
-    protected String docPath;
-    protected String uiPath;
+    private String docPath;
```
Copyright date should be 2023, 2024

Azquelt replied:
Fixed.

tevans78 commented on dev/io.openliberty.microprofile.openapi.2.0.internal.servlet/src/io/openliberty/microprofile/openapi20/internal/servlet/ApplicationServlet.java:
```diff
@@ -32,10 +32,12 @@
 
 import com.ibm.websphere.ras.Tr;
 import com.ibm.websphere.ras.TraceComponent;
+import com.ibm.ws.kernel.productinfo.ProductInfo;
```
Copyright date

Azquelt replied:
Fixed

tevans78 commented on dev/io.openliberty.microprofile.openapi.2.0.internal.servlet/src/io/openliberty/microprofile/openapi20/internal/servlet/ApplicationServlet.java:
```diff
@@ -58,13 +61,16 @@ public void init(ServletConfig config) throws ServletException {
         appRegistryTracker.open();
         modelOperationsTracker = new ServiceTracker<>(bundleContext, OpenAPIModelOperations.class, null);
         modelOperationsTracker.open();
+        versionConfigTracker = new ServiceTracker<>(bundleContext, OpenAPIVersionConfig.class, null);
```
should the init and destroy code for OpenAPIVersionConfig be beta guarded as well?

Azquelt replied:
Yes, we do beta-guard the init code in the [activate method of OpenAPIVersionConfigImpl](https://github.com/OpenLiberty/open-liberty/pull/29877/files#diff-aa40fd44bd64ea333e2485f56da55fb2588a9cd290035c6d3519bce144ab5804R50-R56). There is no destroy code for that class.

Here we're just setting up a tracker for that service. It doesn't actually create a dependency on the service, it just keeps track of whether it's present. We do beta-guard around attempting to retrieve and call the service further down this file.

tevans78 commented on dev/io.openliberty.microprofile.openapi.2.0.internal/src/io/openliberty/microprofile/openapi20/internal/services/OpenAPIVersionConfig.java:
```diff
@@ -0,0 +1,16 @@
+package io.openliberty.microprofile.openapi20.internal.services;
```
Missing copyright header

Azquelt replied:
Bah, I must have done this one before I noticed the project wasn't configured to add them automatically. Fixed.

Azquelt commented:
#build
#libby
#spawn.fullfat.buckets=com.ibm.ws.microprofile.openapi.1.1_fat_tck,com.ibm.ws.microprofile.openapi_fat,com.ibm.ws.microprofile.openapi_fat_tck,com.ibm.ws.openapi.ui_fat,io.openliberty.microprofile.openapi.2.0.internal_fat,io.openliberty.microprofile.openapi.2.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.0.internal_fat_tck,io.openliberty.microprofile.openapi.3.1.internal_fat_tck,io.openliberty.microprofile.openapi.4.0.internal_fat,io.openliberty.microprofile.openapi.4.0.internal_fat_tck,io.openliberty.microprofile.openapi.ui.internal_fat

Azquelt commented:
Failures in the build do not appear to be related to this change.
